### DISCUSSION TITLE
HTTP와 HTTPS의 차이점

### DISCUSSION THREAD

**Discussion by WooGie911** [wanted-9th-3team/frontend-study#21] [프론트 기술 질문 공간] [answered]
HTTPS의 원리를 설명해주세요.

**Comment by skyhanull** [comment]
HTTP(Hyper Text Transfer Protocol)는 서버/클라이언트 모델을 따라 데이터를 주고 받기 위한 프로토콜입니다 
포트는 80번 포트를 사용하며 보안이 적용되지 않는 데이터를 전송합니다
HTTPS는 위의 HTTP에 데이터 암호화가 추가된 프로토콜을 의미합니다 포트는 443번 포트를 사용합니다
즉 HTTP를 SSL(Secure Sockets Layer) 프로토콜 위에서 돌아가도록하여 클라이언트와 서버가 주고받는 텍스트를 암호화합니다.
암호화는 공개키 암호화를 이용하여 하는데
1. 서버는 공개키와 개인키를 만든다.
2. 믿을만한 저장소(CA)와 계약해 두 키를 관리하도록 한다.
3.  CA도 자신만의 공개키와 개인키를 가지고 자체 암호화를 하여 SSL 인증서를 발급한다.
4. 클라이언트의 데이터 요청이 들어오면 서버는 CA에서 만든 SSL 인증서를 보내준다.
5. 클라이언트는 CA의 공개키를 이용해 SSL 인증서를 복호화(암호 해독)한다.
6. SSL 내부의 서버 공개키를 이용해서 암호화해서 인증서를 서버에게 보낸다.
7. 서버는 개인키로 암호화된 인증서를 복호화(암호 해독)하고 다시 암호화 하여 클라이언트에게 보낸다.
 이런식으로 암호화가 이루어집니다.

**Comment by michoball** [comment]
http 는 HTML, css, Js, Img 같은 데이터를 서버와 클라이언트가 주고받기 위해 필요한 모두가 알아들을 수 있는 메시지 입니다. 
정해진 형식이 맞게 서버로 정보를 요청하고 서버는 이에 맞게 응답을 해주는 형식인데요.

기존 http 형식은 서버와 응답을 하는 사이 누군가 악의적인 목적으로 이 데이터를 들여다 본다면 데이터의 원본이 그대로 보여지기 때문에 보안에 매우 취약했습니다. 

그래서 보안을 추가한 형식이 등장하는게 그게 https입니다. 

서버와 클라이언트가 데이터를 주고 받을 때 데이터를 암호화해서 주고 받고 받은 데이터는 암호를 해독하는 키를 이용해 해독해서 데이터가 전송되는 동안에 정보가 새지 않도록 해줍니다. 

암호화한 데이터는 대칭키와 비 대칭 키를 이용해 해독을 하고 중간에서 신뢰도 있게 데이터를 공유하도록 해주는 역할을 하는 CA (Certificate Authority)기업이 존재합니다.

**Comment by WooGie911** [comment]
HTTPS는 HTTP(Hypertext Transfer Protocol)를 기반으로 하는 암호화된 프로토콜입니다. 

HTTPS는 웹 서버와 클라이언트 간의 데이터 전송을 암호화하여 
네트워크 중간자가 전송하는 데이터를 읽거나 조작할 수 없도록 합니다.

HTTP란 서버/클라이언트 모델을 따라 데이터를 주고 받기 위한 프로토콜이며
이 HTTP에는 3가지 문제가 있습니다. 

첫 번째 HTTP 는 평문 통신이기 때문에 도청이 가능하다. 
두 번째 통신 상대를 확인하지 않기 때문에 위장이 가능하다. 
세 번째 완전성을 증명할 수 없기 때문에 변조가 가능하다. 

이 3가지 문제를 해결하기 위해 HTTPS는 SSL(Secure Socket Layer) or TLS(Transport Layer Security)와 같은 프로토콜을 사용하여 공개키/개인키 기반으로 데이터를 암호화하고 있습니다. 

데이터는 암호화되어 전송되기 때문에 임의의 사용자가 데이터를 조회하여도 
원본의 데이터를 보는 것은 불가능하고, 완정성 또한 증명할 수 있습니다.

**Comment by bona373737** [comment]
주요적인 차이점은 보안기능이 추가됬다는 점인데, 이를 설명하기 위해선 프로토콜과 네트워크 통신계층에 대한 이해가 필요할 것 같습니다.

**Comment by WooGie911** [comment, marked as answer]
HTTP  ( 포트번호: 80 )
 - Hyper Text Transfer Protocol → 서버와 클라이언트가 상호간에 데이터를 주고 받기위한 서로가 이해할 수 있는 메시지 규약
 
 - request message header 파트 → request line + request header
   - request line  → GET / index.html  HTTP / 1.1
   - request header  → user-Agent, Accept-Encoding, cookie
   - empty line
   - body ⇒ get요청에는 존재하지 않는다

 - response message
   - Status line ⇒ ex) HTTP/1.1 404 Not Found
   - Headers ⇒ 위와 비슷
   - Body ⇒ 요청에 따른 data
        
 위 http 의 문제 
 - HTTP 는 평문 통신이기 때문에 도청이 가능하다.
 - 통신 상대를 확인하지 않기 때문에 위장이 가능하다.
 - 완전성을 증명할 수 없기 때문에 변조가 가능하다.

HTTPS ( 포트번호: 443 )
 - 데이터를 주고 받을 때 암호화해서 주고받는 방식
 
 - 노출이 되어도 괜찮은 데이터는 HTTP를 이용해도 되지만, 구글에서는 SEO를 위해 HTTPS를 권장
  → HTTP 란 서버와 클라이언트가 상호간에 데이터를 주고 받기 위한 서로가 이해할 수 있는 메시지 규약.

결론 : 
 HTTP를 사용했을 때는 데이터가 전송되는 중간에 도청이 가능하고, 
위장 변조가 가능해서 중요한 개인정보가 노출될 수 있는 위험이 있습니다. 
 그래서 기존 HTTP에 보안을 강화해서 데이터를 암호화해서 주고 받는 형식을 만들었고 이게 HTTPS입니다. 


참고 : 얄코 [https://youtu.be/H6lpFRpyl14](https://youtu.be/H6lpFRpyl14)
